src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Serialization/TagHelperDescriptorJsonConverter.cs

```diff
@@ -24,10 +26,35 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
                 return null;
             }
 
+            // Try reading the optional hashcode
+            var hashWasRead = reader.TryReadNextProperty(RazorSerializationConstants.HashCodePropertyName, out var hashObj);
+            var hash = hashObj as int?;
+            if (!DisableCachingForTesting &&
+                hashWasRead &&
+                hash.HasValue &&
+                TagHelperDescriptorCache.TryGetDescriptor(hash.Value, out var descriptor))
+            {
+                ReadToEndOfCurrentObject(reader);
+                return descriptor;
+            }
+
             // Required tokens (order matters)
-            var descriptorKind = reader.ReadNextStringProperty(nameof(TagHelperDescriptor.Kind));
-            var typeName = reader.ReadNextStringProperty(nameof(TagHelperDescriptor.Name));
-            var assemblyName = reader.ReadNextStringProperty(nameof(TagHelperDescriptor.AssemblyName));
+            if (!reader.TryReadNextProperty(nameof(TagHelperDescriptor.Kind), out var descriptorKindObj) ||
+                !(descriptorKindObj is string descriptorKind))
+            {
+                return default;
+            }
+            if (!reader.TryReadNextProperty(nameof(TagHelperDescriptor.Name), out var typeNameObj) ||
+                !(typeNameObj is string typeName))
```

Comment:
You could make TryReadNextProperty a generic and it can do this cast

Reply:
Yeah trying that, just running into a couple issues with how JsonReader deserializes integers by default (Int64) which causes casting to fail to int32 without special handling. Updating...

Reply:
https://github.com/dotnet/aspnetcore-tooling/pull/2307/commits/66e8b7eb9d1c3f2bd309a525f0f4bd3a91dc65d4